skbio/stats/distance/_pairwise_distances.py

```diff
@@ -0,0 +1,446 @@
+        >>> len(pwdist)
+        4
+
+        Note that there are three "within" distances and one "between"
+        distance.
+
+        """
+        return len(self._distances) + len(self.ids)
+
+    @experimental(as_of="0.4.2-dev")
+    def __iter__(self):
+        """Iterate over "within" and "between" ID pairs in no particular order.
+
+        Yields "within" (ID to same ID) and "between" (ID to different ID)
+        ID pairs. Does not yield redundant distances (e.g., only one of A vs.
+        B or B vs. A will be yielded; which ordering is yielded is not
+        guaranteed).
+
+        There is no guaranteed order to how ID pairs are yielded.
+
+        Yields
+        ------
+        tuple
+            Each "within" and "between" ID pair present in this
+            ``PairwiseDistances``.
+
+        Examples
+        --------
+        >>> from skbio.stats.distance import PairwiseDistances
+        >>> pwdist = PairwiseDistances({('A', 'B'): 0.1, ('C', 'C'): 0.0})
+        >>> for id_pair in pwdist:
+        ...     # do something with `id_pair`, such as `pwdist[id_pair]`
+        ...     pass
+
+        """
+        yield from self._distances
+        for id_ in self.ids:
+            yield id_, id_
+
+    @experimental(as_of="0.4.2-dev")
+    def __getitem__(self, id_pair):
+        """Retrieve the distance between a pair of IDs.
+
+        Supports retrieving both "within" and "between" distances.
+
+        Parameters
+        ----------
+        id_pair : tuple
+            Pair of IDs whose distance will be retrieved. Either ordering of
+            IDs is supported, such that A vs. B and B vs. A will return the
+            same distance.
+
+        Returns
+        -------
+        float
+            Distance between IDs in `id_pair`.
+
+        Raises
+        ------
+        TypeError
+            If `id_pair` isn't a tuple of exactly two strings.
+        KeyError
+            If the pair of IDs isn't present in this ``PairwiseDistances``.
+
+        Examples
+        --------
+        >>> from skbio.stats.distance import PairwiseDistances
+        >>> pwdist = PairwiseDistances({('A', 'B'): 0.1, ('C', 'C'): 0.0})
+        >>> pwdist['A', 'B']
+        0.1
+        >>> pwdist['B', 'A']
+        0.1
+        >>> pwdist['A', 'A']
+        0.0
+        >>> pwdist['B', 'B']
+        0.0
+        >>> pwdist['C', 'C']
+        0.0
+
+        """
+        if is_id_pair(id_pair):
+            if id_pair in self._distances:
+                return self._distances[id_pair]
+            elif (id_pair[1], id_pair[0]) in self._distances:
+                return self._distances[id_pair[1], id_pair[0]]
+            elif id_pair[0] == id_pair[1] and id_pair[0] in self.ids:
+                return 0.0
+            else:
+                raise KeyError(id_pair)
+        else:
+            raise TypeError(
+                "`id_pair` must be a tuple of exactly two strings, not %r"
```

Comment:
For the block decomposition method, there is a large benefit in using integer IDs as the IDs map directly into the counts array for slicing.